db/fetcher/indexer_iterators.go

```diff
@@ -0,0 +1,462 @@
+	}
+	key, err := core.NewIndexDataStoreKey(res.Key)
+	if err != nil {
+		return core.IndexDataStoreKey{}, false, err
+	}
+	return key, true, nil
+}
+
+func (i queryResultIterator) Close() error {
+	return i.resultIter.Close()
+}
+
+type eqIndexIterator struct {
+	queryResultIterator
+	indexKey  core.IndexDataStoreKey
+	filterVal []byte
+	execInfo  *ExecInfo
+}
+
+func (i *eqIndexIterator) Init(ctx context.Context, store datastore.DSReaderWriter) error {
+	i.indexKey.FieldValues = [][]byte{i.filterVal}
+	resultIter, err := store.Query(ctx, query.Query{
+		Prefix:   i.indexKey.ToString(),
+		KeysOnly: true,
+	})
+	if err != nil {
+		return err
+	}
+	i.resultIter = resultIter
+	return nil
+}
+
+func (i *eqIndexIterator) Next() (core.IndexDataStoreKey, bool, error) {
+	key, hasValue, err := i.queryResultIterator.Next()
+	if hasValue {
+		i.execInfo.IndexesFetched++
+	}
+	return key, hasValue, err
+}
+
+type inIndexIterator struct {
+	eqIndexIterator
+	filterValues [][]byte
+	nextValIndex int
+	ctx          context.Context
+	store        datastore.DSReaderWriter
+	hasIterator  bool
+}
+
+func newInIndexIterator(
+	indexKey core.IndexDataStoreKey,
+	filterValues [][]byte,
+	execInfo *ExecInfo,
+) *inIndexIterator {
+	return &inIndexIterator{
+		eqIndexIterator: eqIndexIterator{
+			indexKey: indexKey,
+			execInfo: execInfo,
+		},
+		filterValues: filterValues,
+	}
+}
+
+func (i *inIndexIterator) nextIterator() (bool, error) {
+	if i.nextValIndex > 0 {
+		err := i.eqIndexIterator.Close()
+		if err != nil {
+			return false, err
+		}
+	}
+
+	if i.nextValIndex >= len(i.filterValues) {
+		return false, nil
+	}
+
+	i.filterVal = i.filterValues[i.nextValIndex]
+	err := i.eqIndexIterator.Init(i.ctx, i.store)
+	if err != nil {
+		return false, err
+	}
+	i.nextValIndex++
+	return true, nil
+}
+
+func (i *inIndexIterator) Init(ctx context.Context, store datastore.DSReaderWriter) error {
+	i.ctx = ctx
+	i.store = store
+	var err error
+	i.hasIterator, err = i.nextIterator()
+	return err
+}
+
+func (i *inIndexIterator) Next() (core.IndexDataStoreKey, bool, error) {
+	for i.hasIterator {
+		key, hasValue, err := i.eqIndexIterator.Next()
+		if err != nil {
+			return core.IndexDataStoreKey{}, false, err
+		}
+		if !hasValue {
+			i.hasIterator, err = i.nextIterator()
+			if err != nil {
+				return core.IndexDataStoreKey{}, false, err
+			}
+			continue
+		}
+		return key, true, nil
+	}
+	return core.IndexDataStoreKey{}, false, nil
+}
+
+func (i *inIndexIterator) Close() error {
+	return nil
+}
+
+type errorCheckingFilter struct {
+	matcher indexMatcher
+	err     error
+}
+
+func (f *errorCheckingFilter) Filter(e query.Entry) bool {
+	if f.err != nil {
+		return false
+	}
+	indexKey, err := core.NewIndexDataStoreKey(e.Key)
+	if err != nil {
+		f.err = err
+		return false
+	}
+	res, err := f.matcher.Match(indexKey)
+	if err != nil {
+		f.err = err
+		return false
+	}
+	return res
+}
+
+type execInfoIndexMatcherDecorator struct {
```

Comment:
todo: Sorry to be a bother, whilst this name is longer, I still have no idea what it is or what it is for - could you please document it?

Reply:
done